File: src/utils/bounding_box.go
package utils

import (
	pb "github.com/bbanez/catan/gen/proto_utils"
)

type BoundingBox struct {
	Size     *Size
	Position *Point
	HalfSize *Size
	Edges    *Cubic
}

func NewBoundingBox(size *Size, position *Point) *BoundingBox {
	bb := &BoundingBox{
		Size:     size,
		Position: position,
		HalfSize: NewSize(size.Width/2, size.Height/2),
		Edges:    NewCubic(0.0, 0.0, 0.0, 0.0),
	}
	bb.update()
	return bb
}

func NewEmptyBoundingBox() *BoundingBox {
	return NewBoundingBox(NewSize(0, 0), NewPoint(0, 0))
}

func BoundingBoxFromProto(bb *pb.BoundingBox) *BoundingBox {
	return NewBoundingBox(SizeFromProto(bb.Size), PointFromProto(bb.Position))
}

func BoundingBoxToProto(bb *BoundingBox) *pb.BoundingBox {
	return &pb.BoundingBox{
		Position: PointToProto(bb.Position),
		Size:     SizeToProto(bb.Size),
	}
}

func (bb *BoundingBox) ToProto() *pb.BoundingBox {
	return BoundingBoxToProto(bb)
}

func (bb *BoundingBox) update() {
	bb.HalfSize.Width = bb.Size.Width / 2
	bb.HalfSize.Height = bb.Size.Height / 2
	bb.Edges.A = bb.Position.Y - bb.HalfSize.Height
	bb.Edges.B = bb.Position.X + bb.HalfSize.Width
	bb.Edges.C = bb.Position.Y + bb.HalfSize.Height
	bb.Edges.D = bb.Position.X - bb.HalfSize.Width
}

func (bb *BoundingBox) SetSize(size *Size) {
	bb.Size.Width = size.Width
	bb.Size.Height = size.Height
	bb.update()
}

func (bb *BoundingBox) SetPosition(position *Point) {
	bb.Position.X = position.X
	bb.Position.Y = position.Y
	bb.update()
}

func (bb *BoundingBox) IsPointInside(point *Point) bool {
	return CollisionWithBbAndPoint(bb.Edges, point)
}

func (bb *BoundingBox) DoesIntersect(bb2 *Cubic) bool {
	return CollisionWithBbAndBb(bb.Edges, bb2)
}

File: frontend/src/proto/utils/bounding-box.ts
import { create } from '@bufbuild/protobuf';
import {
    BoundingBoxSchema,
    type BoundingBox as ProtoBoundingBox,
} from '@proto/proto_utils/bounding_box_pb';
import { Point } from '@root/proto/utils/point';
import { Size } from '@root/proto/utils/size';
import { Cubic } from '@root/proto/utils/cubic';
import { Collision } from '@root/game/utils/collision';

export class BoundingBox {
    halfSize = new Size(0, 0);
    edges = new Cubic(0, 0, 0, 0);

    constructor(
        public position: Point,
        public size: Size,
    ) {
        this.update();
    }

    private update() {
        this.halfSize.width = this.size.width / 2;
        this.halfSize.height = this.size.height / 2;
        this.edges.a = this.position.y - this.halfSize.height;
        this.edges.b = this.position.x + this.halfSize.width;
        this.edges.c = this.position.y + this.halfSize.height;
        this.edges.d = this.position.x - this.halfSize.width;
    }

    copy(): BoundingBox {
        return new BoundingBox(this.position.copy(), this.size.copy());
    }

    setSize(width: number, height: number) {
        this.size.width = width;
        this.size.height = height;
        this.update();
    }

    setPosition(x: number, y: number) {
        this.position.x = x;
        this.position.y = y;
        this.update();
    }

    isPointInside(x: number, y: number): boolean {
        return Collision.withBoxAndPoint(this.edges, new Point(x, y));
    }

    doesIntersect(other: Cubic): boolean {
        return Collision.withBoxAndBox(this.edges, other);
    }

    toProto(): ProtoBoundingBox {
        return create(BoundingBoxSchema, {
            position: Point.toProto(this.position),
            size: Size.toProto(this.size),
        });
    }

    static fromProto(proto: ProtoBoundingBox): BoundingBox {
        if (!proto.position) {
            throw new Error('Property "position" is required');
        }
        if (!proto.size) {
            throw new Error('Property "size" is required');
        }
        return new BoundingBox(
            Point.fromProto(proto.position),
            Size.fromProto(proto.size),
        );
    }

    static newEmpty(): BoundingBox {
        return new BoundingBox(new Point(0, 0), new Size(0, 0));
    }
}
